plugins/embed-optimizer/class-embed-optimizer-tag-visitor.php

```diff
@@ -119,6 +158,80 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 			$this->added_lazy_script = true;
 		}
 
-		return true;
+		/*
+		 * At this point the tag is a figure.wp-block-embed, and we can return false because this does not need to be
+		 * measured and stored in URL Metrics. Only the child div.wp-block-embed__wrapper tag is measured and stored
+		 * so that this visitor can look up the height to set as a min-height on the figure.wp-block-embed. For more
+		 * information on what the return values mean for tag visitors, see <https://github.com/WordPress/performance/issues/1342>.
+		 */
+		return false;
+	}
+
+	/**
+	 * Gets the XPath for the embed wrapper.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context.
+	 * @return string XPath.
+	 */
+	private function get_embed_wrapper_xpath( OD_Tag_Visitor_Context $context ): string {
+		return $context->processor->get_xpath() . '/*[1][self::DIV]';
+	}
+
+	/**
+	 * Reduces layout shifts.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context.
+	 */
+	private function reduce_layout_shifts( OD_Tag_Visitor_Context $context ): void {
+		$processor           = $context->processor;
+		$embed_wrapper_xpath = $this->get_embed_wrapper_xpath( $context );
+
+		/**
+		 * Array of tuples of groups and their minimum heights keyed by the minimum viewport width.
+		 *
+		 * @var array<int, array{OD_URL_Metric_Group, int}> $group_minimum_heights
+		 */
+		$group_minimum_heights = array();
+
+		$denormalized_elements = $context->url_metric_group_collection->get_all_denormalized_elements()[ $embed_wrapper_xpath ] ?? array();
+		foreach ( $denormalized_elements as list( $group, $url_metric, $element ) ) {
```

Comment:
See note below on the `get_all_denormalized_elements()` method that I'd like to instead have `$element` be an `OD_Element` class instance which has a `$url_metric` property to access the `OD_URL_Metric` instance if is a part of. In the same way, the `OD_URL_Metric` class can have a `$group` property to indicate the `OD_URL_Metric_Group` it is a part of. This would eliminate the need to pass back an array of tuples and instead. So instead, to get the `$group` this code could do `$element->url_metric->group`.

Reply:
> See note below on the `get_all_denormalized_elements()` method that I'd like to instead have `$element` be an `OD_Element` class instance which has a `$url_metric` property to access the `OD_URL_Metric` instance if is a part of. In the same way, the `OD_URL_Metric` class can have a `$group` property to indicate the `OD_URL_Metric_Group` it is a part of. This would eliminate the need to pass back an array of tuples and instead. So instead, to get the `$group` this code could do `$element->url_metric->group`.
> 
> This I want to do in a follow-up PR.

Reply:
See https://github.com/WordPress/performance/pull/1585